src/encoding.rs

```diff
@@ -1680,11 +1679,10 @@ mod test {
 
     #[test]
     fn varint_overflow() {
-        let mut u64_max_plus_one: &[u8] =
-            &[0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0x02];
+        let u64_max_plus_one: &[u8] = &[0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0x02];
 
-        decode_varint(&mut u64_max_plus_one).expect_err("decoding u64::MAX + 1 succeeded");
-        decode_varint_slow(&mut u64_max_plus_one)
+        decode_varint(&mut &*u64_max_plus_one).expect_err("decoding u64::MAX + 1 succeeded");
+        decode_varint_slow(&mut &*u64_max_plus_one)
```

Comment:
I suggest improving code readability here. The `&*` syntax doesn't explain what is going on. I would do something like this to explicitly create a separated buffer. I think the extra allocations are not a problem in test code.

```suggestion

        let mut buffer = Bytes::copy_from_slice(u64_max_plus_one);
        decode_varint_slow(&mut buffer).expect_err("slow decoding u64::MAX + 1 succeeded");
```

Reply:
I wrote `&*` here because it is spelled similarly nearby. The explicit spelling of this action is `.as_slice()`, which I'm happy to do.

Reply:
I made every occurrence of `&*` reborrows explicit; if you would prefer I can scope that down but it seems like an improvement.

Reply:
> I wrote `&*` here because it is spelled similarly nearby. The explicit spelling of this action is `.as_slice()`, which I'm happy to do.

Sorry, I was not aware that the other code already uses `&*` syntax. Your first commit does the correct thing: make the tests uniform.


> I made every occurrence of `&*` reborrows explicit; if you would prefer I can scope that down but it seems like an improvement.

I think the new syntax (`as_slice` and such) don't convey the intent of the call. What the code is doing is creating a fresh `Buf` for each call to decode. So I am thinking of words like copy or clone. Maybe a new function like `fn decouple(&[u8]) -> impl Buf` or something like that. Do you have ideas?

Reply:
When the data is in a `Vec`, it does not implement `Buf`. To get a `Buf`, we need to get it as a primitive `&[u8]` slice, which **_does_** implement the trait. (From the [`bytes` documentation](https://docs.rs/bytes/latest/bytes/buf/trait.Buf.html): "The simplest `Buf` is a `&[u8]`.")

We can do this either by using the existence of the `Deref<Target = [u8]>` trait on the vec to re-borrow the value tersely with `&*`, or you can get it explicitly with `Vec::as_slice`. In my opinion the more verbose way conveys the intent of the call perfectly, and i believe the general confusion around all of this (including the extistence of the original bug) is evidence that it's needed.

Reply:
What might make the most sense is to just use the `Bytes::from_static` and then clone that. But I don't think it matters much. As for `&*` vs `as_slice` as_slice is much better imo.